Mathlib/Algebra/Order/Hom/Quantale.lean

```diff
@@ -0,0 +1,218 @@
+    α →ₙ*q β := { (f : α →ₙ* β) with map_sSup' := sSupHomClass.map_sSup f }
+
+/-- Any type satisfying `QuantaleHomClass` can be cast into `QuantaleHom` via
+  `QuantaleHomClass.toQuantaleHom`. -/
+@[to_additive "Any type satisfying `AddQuantaleHomClass` can be cast into `AddQuantaleHom` via
+  `AddQuantaleHomClass.toAddQuantaleHom`."]
+instance [MulHomClass F α β] [sSupHomClass F α β] : CoeTC F (α →ₙ*q β) :=
+  ⟨toQuantaleHom⟩
+
+end Quantale
+
+namespace QuantaleHom
+
+variable [Semigroup α] [Semigroup β] [Semigroup γ] [Semigroup δ]
+  [CompleteLattice α] [CompleteLattice β] [CompleteLattice γ] [CompleteLattice δ]
+  {f g : α →ₙ*q β}
+
+@[to_additive]
+instance : FunLike (α →ₙ*q β) α β where
+  coe f := f.toFun
+  coe_injective' f g h := by
+    obtain ⟨⟨ _ ⟩, _⟩ := f
+    obtain ⟨⟨ _ ⟩, _⟩ := g
+    congr
+
+@[to_additive]
+instance : MulHomClass (α →ₙ*q β) α β where
+  map_mul f _ _ := f.map_mul' _ _
+
+@[to_additive]
+instance : sSupHomClass (α →ₙ*q β) α β where
+  map_sSup f := f.map_sSup'
+
+@[to_additive]
+instance : OrderHomClass (α →ₙ*q β) α β where
+  map_rel f := by apply (f : OrderHom α β).monotone'
+
+@[to_additive (attr := ext)]
+theorem ext (h : ∀ a, f a = g a) : f = g :=
+  DFunLike.ext f g h
+
+@[to_additive]
+theorem toFun_eq_coe (f : α →ₙ*q β) : f.toFun = (f : α → β) := rfl
+
+@[to_additive (attr := simp)]
+theorem coe_mk (f : α →ₙ* β) (h) : (QuantaleHom.mk f h : α → β) = f := rfl
+
+@[to_additive (attr := simp)]
+theorem mk_coe (f : α →ₙ*q β) (h) : QuantaleHom.mk (f : α →ₙ* β) h = f := rfl
+
+section Id
+
+variable (α)
+
+/-- The identity map as a quantale homomorphism. -/
+@[to_additive "The identity map as an additive quantale homomorphism."]
+protected def id : α →ₙ*q α where
+  toFun x := x
+  map_mul' _ _ := rfl
+  map_sSup' _ := by simp_all only [Set.image_id']
+
+@[to_additive (attr := simp)]
+theorem coe_id : ⇑(QuantaleHom.id α) = id := rfl
+
+@[to_additive]
+instance : Inhabited (α →ₙ*q α) := ⟨QuantaleHom.id α⟩
+
+end Id
+
+section Comp
+
+/-- Composition of `QuantaleHom`s as a `QuantaleHom`. -/
+@[to_additive "Composition of `AddQuantaleHom`s as an `AddQuantaleHom`"]
+def comp (f : β →ₙ*q γ) (g : α →ₙ*q β) : α →ₙ*q γ :=
+  {f.toMulHom.comp (g : α →ₙ* β), f.tosSupHom.comp (g : sSupHom α β) with }
+
+@[to_additive (attr := simp)]
+theorem coe_comp (f : β →ₙ*q γ) (g : α →ₙ*q β) : (f.comp g : α → γ) = f ∘ g :=
+  rfl
+
+@[to_additive (attr := simp)]
+theorem comp_apply (f : β →ₙ*q γ) (g : α →ₙ*q β) (a : α) : (f.comp g) a = f (g a) :=
+  rfl
+
+@[to_additive (attr := simp)]
+theorem comp_assoc (f : γ →ₙ*q δ) (g : β →ₙ*q γ) (h : α →ₙ*q β) :
+    (f.comp g).comp h = f.comp (g.comp h) :=
+  rfl
+
+@[to_additive (attr := simp)]
+theorem comp_id (f : α →ₙ*q β) : f.comp (QuantaleHom.id α) = f :=
+  rfl
+
+@[to_additive (attr := simp)]
+theorem id_comp (f : α →ₙ*q β) : (QuantaleHom.id β).comp f = f :=
+  rfl
+
+end Comp
+
+section Bot
+
+/-- `⊥` is the quantale homomorphism sending all elements to `⊥`. -/
+
+@[to_additive]
```

Comment:
Can you either give a docstring to both the additive and multiplicative instances, or to neither of them?